File: week1/pset1/cash/cash.c
// Greedy algorithm to calculate minimum amount of coins for owed cents user inputs

#include <cs50.h>
#include <stdio.h>

// Definitions of values of coins
#define QUARTER_VALUE 25
#define DIME_VALUE 10
#define NICKEL_VALUE 5
#define PENNY_VALUE 1

int get_cents(void);
int calculate_quarters(int cents);
int calculate_dimes(int cents);
int calculate_nickels(int cents);
int calculate_pennies(int cents);

int main(void)
{
    // Prompt for how many cents the customer is owed
    int cents = get_cents();

    // Calculate the number of quarters to give the customer
    int quarters = calculate_quarters(cents);
    cents = cents - quarters * QUARTER_VALUE;

    // Calculate the number of dimes to give the customer
    int dimes = calculate_dimes(cents);
    cents = cents - dimes * DIME_VALUE;

    // Calculate the number of nickels to give the customer
    int nickels = calculate_nickels(cents);
    cents = cents - nickels * NICKEL_VALUE;

    // Calculate the number of pennies to give the customer
    int pennies = calculate_pennies(cents);
    cents = cents - pennies * PENNY_VALUE;

    // Sum coins
    int coins = quarters + dimes + nickels + pennies;

    // Print total number of coins to give the customer
    printf("%i\n", coins);
}

// Prompt user for cents, check if 0 or positive
int get_cents(void)
{
    int n;
    do
    {
        n = get_int("Change owed: ");
    }
    while (n < 0);
    return n;
}

// Calculate the amount of quarters
int calculate_quarters(int cents)
{
    // Division of two ints will truncate the decimal part, the cents that can't be even one coin of this value
    return cents / QUARTER_VALUE;
}

// Calculate the amount of dimes
int calculate_dimes(int cents)
{
    return cents / DIME_VALUE;
}

// Calculate the amount of nickels
int calculate_nickels(int cents)
{
    return cents / NICKEL_VALUE;
}

// Calculate the amount of pennies
int calculate_pennies(int cents)
{
    return cents / PENNY_VALUE;
}

File: week6/pset6/sentimental-cash/cash.py
# Greedy algorithm to calculate minimum amount of coins for owed change user inputs

from cs50 import get_float

# Definitions of values of coins
QUARTER, DIME, NICKEL, PENNY = 25, 10, 5, 1
coins_list = [QUARTER, DIME, NICKEL, PENNY]


def main():
    # Prompt for how much change the customer is owed
    change = get_change()

    # Print total number of coins to give the customer
    print(calculate_coins(change))


# Prompt user for change, check if 0 or positive
def get_change():
    while True:
        n = get_float("Change owed: ")
        if n >= 0:
            return n


# Calculate total number of coins to give the user
def calculate_coins(change):
    # Get only the integer part of change, each unit is equivalent to 4 quarters
    coins = int(change) * 4

    # Update change to be its cents part
    change = int((change - int(change)) * 100)

    for coin in coins_list:
        # Divide and truncate change by every coin value to get the amount of that coin
        coins += change // coin
        # Update change to be the value that can't fit in that coin's value
        change %= coin

    return coins


main()
